closes #51 closes #74 closes #75 documentation increase unit test coverage for derive params growth age

Minlei0201 opened the pull request:
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiralpeds (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiralpeds site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the Reference page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

Minlei0201 commented:
Several notes/questions on the unit tests:
1). Test cases for patients < 2 years old not added yet, since there is no P95 and sigma in metadata (which are required to derive extended BMI SDS and percentiles).
2). Test cases for out of bound age not added yet: the expected result for such case should be missing SDS and percentiles, correct?
3). Test cases for missing anthropometric values not added yet: the expected result for such case should be missing SDS and percentiles, correct?
4). Is there any differences in handling length vs. height? If so, such test cases will be included as well.

rossfarrugia commented:
@Minlei0201 the R CMD CI check fail looks to be as some of the tests expected and actual don't match (see https://github.com/pharmaverse/admiralpeds/actions/runs/9840494918/job/27164859357?pr=67). Maybe its because this update needs to be merged first and then hopefully they'll match: https://github.com/pharmaverse/admiralpeds/pull/60

Fanny-Gautier commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,41 +1,291 @@
+  )
+
+  actual <- derive_params_growth_age(
+    dataset = vs_data,
```
Should the new argument `by_vars` be added there ?

rossfarrugia replied:
good point! same is true for https://github.com/pharmaverse/admiralpeds/pull/54/files. maybe @zdz2101 could make a commit to both these branches to update this after we merge his PR: https://github.com/pharmaverse/admiralpeds/pull/60, as looks like that one is going to need to merge first.

Minlei0201 replied:
Thanks for catching! `by_vars` added to all test codes.

Fanny-Gautier commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,41 +1,291 @@
+  )
+
+  actual <- derive_params_growth_age(
+    dataset = vs_data,
```
Should the new argument `by_vars` be added there ?

Minlei0201 replied:
Thanks for catching! Yes, `by_vars` added to all test codes.

Fanny-Gautier commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,41 +1,291 @@
+  )
+
+  actual <- derive_params_growth_age(
+    dataset = vs_data,
```
Should the new argument `by_vars` be added there ?

Minlei0201 replied:
Thanks Fanny! Yes, `by_vars` added to all test cases.

Fanny-Gautier commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,41 +1,291 @@
+  )
+
+  actual <- derive_params_growth_age(
+    dataset = vs_data,
```
Should the new argument `by_vars` be added there ?

Minlei0201 replied:
Thanks Fanny! `by_vars` added to all tests.

Minlei0201 commented:
> @Minlei0201 the R CMD CI check fail looks to be as some of the tests expected and actual don't match (see https://github.com/pharmaverse/admiralpeds/actions/runs/9840494918/job/27164859357?pr=67). Maybe its because this update needs to be merged first and then hopefully they'll match: #60

Hi @rossfarrugia , do you mean creating a PR and merge #60 to #51 ? Currently there are some conflict in the unit testing file before I can merge. @zdz2101 , maybe you can remove this file so that I can merge your branch into mine?
https://github.com/pharmaverse/admiralpeds/blob/59-derive_params_growth_age-add-interpolation/tests/testthat/test-derive_params_growth_age.R

rossfarrugia commented:
> Hi @rossfarrugia , do you mean creating a PR and merge #60 to #51 ? Currently there are some conflict in the unit testing file before I can merge. @zdz2101 , maybe you can remove this file so that I can merge your branch into mine? https://github.com/pharmaverse/admiralpeds/blob/59-derive_params_growth_age-add-interpolation/tests/testthat/test-derive_params_growth_age.R

@Minlei0201 no, don't merge Zelos's branch yet as its still under review. My thoughts were let's get his reviewed/approved and then merged to main, and then you could run a `git merge main` in this branch. It will lead to merge conflicts for the unit test script but then you could make a commit to blend yours and Zelos's updates for the unit tests. I think some of the tests he added likely complement your checks, so i think there is value in combining unless you think all your tests already cover everything Zelos added to his branch.

Minlei0201 commented:
> > Hi @rossfarrugia , do you mean creating a PR and merge #60 to #51 ? Currently there are some conflict in the unit testing file before I can merge. @zdz2101 , maybe you can remove this file so that I can merge your branch into mine? https://github.com/pharmaverse/admiralpeds/blob/59-derive_params_growth_age-add-interpolation/tests/testthat/test-derive_params_growth_age.R
> 
> @Minlei0201 no, don't merge Zelos's branch yet as its still under review. My thoughts were let's get his reviewed/approved and then merged to main, and then you could run a `git merge main` in this branch. It will lead to merge conflicts for the unit test script but then you could make a commit to blend yours and Zelos's updates for the unit tests. I think some of the tests he added likely complement your checks, so i think there is value in combining unless you think all your tests already cover everything Zelos added to his branch.

Hi @rossfarrugia , I see. So the plan for me is to incorporate the tests from Zelos' branch, inform @zdz2101 to incorporate the most recent unit test code from #51 (to avoid conflict in the next step), and merge main into #51 after Zelos' #60 merge into main. Does that sound reasonable?

zdz2101 commented:
> > > Hi @rossfarrugia , do you mean creating a PR and merge #60 to #51 ? Currently there are some conflict in the unit testing file before I can merge. @zdz2101 , maybe you can remove this file so that I can merge your branch into mine? https://github.com/pharmaverse/admiralpeds/blob/59-derive_params_growth_age-add-interpolation/tests/testthat/test-derive_params_growth_age.R
> > 
> > 
> > @Minlei0201 no, don't merge Zelos's branch yet as its still under review. My thoughts were let's get his reviewed/approved and then merged to main, and then you could run a `git merge main` in this branch. It will lead to merge conflicts for the unit test script but then you could make a commit to blend yours and Zelos's updates for the unit tests. I think some of the tests he added likely complement your checks, so i think there is value in combining unless you think all your tests already cover everything Zelos added to his branch.
> 
> Hi @rossfarrugia , I see. So the plan for me is to incorporate the tests from Zelos' branch, inform @zdz2101 to incorporate the most recent unit test code from #51 (to avoid conflict in the next step), and merge main into #51 after Zelos' #60 merge into main. Does that sound reasonable?

This would work, might even be easier to instead of targeting the `main` branch to merge you can target `59-derive_params_growth_age-add-interpolation` , the branch containing the new function changes. We can merge yours in beforehand, then once PR #60 is ready to merge it'll also already contain your unit-tests

zdz2101 commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,41 +1,398 @@
 # derive_params_growth_age ----
+library(dplyr)
```
```suggestion
```

Per [textbook](https://r-pkgs.org/testing-basics.html#test-organisation)  you'll want to avoid calling libraries in unit-test suite, if you need to use {pkgname} functions just use `pkgname::xxx()` format, demonstrated below with the `tibble::tribble()` call. Although it's ideal to this with all outside-functions, I think `dplyr` ones are so commonly used we can use the `importFrom` in the `admiralpeds-package.R` to avoid excessive use

zdz2101 commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,41 +1,398 @@
+    mutate(AGE = AGECUR_c, AGEU = AGEU_c) %>%
+    select(-ends_with("_c"), -AGECUR)
+
+  vs_data_meta <- vs_data_age %>% inner_join(meta, by = c("SEX", "AGE", "AGEU"))
```
I suspect the `inner_join()` call here is why you needed to put `library(dplyr)` at the top, you can either do `dplyr::inner_join()` or just put `inner_join()` into the `importFrom`

Minlei0201 replied:
Thanks @zdz2101 ! Will update in the next commit.

zdz2101 reviewed (commented):
@Minlei0201 If you take a look at the "new" test suite in #60 you'll see I've added multiple visits _per_ patient here and there to demonstrate clearly how the `by_vars` argument needs to be used, if you want to just throw in a couple more lines for a particular subject, it would closely match mine

Minlei0201 commented:
> @Minlei0201 If you take a look at the "new" test suite in #60 you'll see I've added multiple visits _per_ patient here and there to demonstrate clearly how the `by_vars` argument needs to be used, if you want to just throw in a couple more lines for a particular subject, it would closely match mine

Hi @zdz2101 , I see, makes sense. I will incorporate these tests to my testing tomorrow.

Minlei0201 commented:
> > > > Hi @rossfarrugia , do you mean creating a PR and merge #60 to #51 ? Currently there are some conflict in the unit testing file before I can merge. @zdz2101 , maybe you can remove this file so that I can merge your branch into mine? https://github.com/pharmaverse/admiralpeds/blob/59-derive_params_growth_age-add-interpolation/tests/testthat/test-derive_params_growth_age.R
> > > 
> > > 
> > > @Minlei0201 no, don't merge Zelos's branch yet as its still under review. My thoughts were let's get his reviewed/approved and then merged to main, and then you could run a `git merge main` in this branch. It will lead to merge conflicts for the unit test script but then you could make a commit to blend yours and Zelos's updates for the unit tests. I think some of the tests he added likely complement your checks, so i think there is value in combining unless you think all your tests already cover everything Zelos added to his branch.
> > 
> > 
> > Hi @rossfarrugia , I see. So the plan for me is to incorporate the tests from Zelos' branch, inform @zdz2101 to incorporate the most recent unit test code from #51 (to avoid conflict in the next step), and merge main into #51 after Zelos' #60 merge into main. Does that sound reasonable?
> 
> This would work, might even be easier to instead of targeting the `main` branch to merge you can target `59-derive_params_growth_age-add-interpolation` , the branch containing the new function changes. We can merge yours in beforehand, then once PR #60 is ready to merge it'll also already contain your unit-tests

Hi @zdz2101 , sounds good. Could you let me know when #59 is ready to merge?

rossfarrugia commented:
@zdz2101 @Minlei0201 i merged https://github.com/pharmaverse/admiralpeds/pull/60 as me and @Fanny-Gautier had approved it today and that should make easier with this one, as you can merge main to here and just focus on resolving the unit test program merge conflicts here in this one branch.

Minlei0201 commented:
Hi @rossfarrugia , @zdz2101 , I've merged main into #51 , and updated unit test to include byvar test cases. Currently there are a few mismatches/questions I'd like to discuss with you:
1. The FL code doesn't produce SDS and percentile when patient's age is in month (see example in Test 1). Per our previous discussion, if patient's age is reported in months, it will be converted and rounded to the nearest age in days by multiplying 30.4375, and then apply the WHO/interpolated CDC data, is that correct?

2. For out of bound age and/or missing anthropometric measures, if the user requests both SDS and percentile, to my understanding, the function is supposed to return two records with missing values (i.e., one for SDS and one for percentile). However, currently the function only returns one record with missing value (see Test 6 and 7). Is that what you expect?

Minlei0201 commented:
> @zdz2101 please do add anything, but my understanding of the 2 questions is:
> 
> 1. No age conversion happens in the function. If you supply age in months in the data then you should also supply age in months in the metadata. In our template and vignette we make the metadata consistently in days via conversion and interpolation for CDC, and then we derive current age in days. Some other company could choose to only supply metadata in months and therefore derive current age in months, and the function should still work (always matching to the nearest when falling in between).
> 2. I would agree to have both parameters created as missing would make sense to me. Maybe this is an update needed to the function?

Hi @rossfarrugia , thanks for your clarification! 
1. I've changed the age unit in days for the majority of tests, and added one test case for age in months. 
2. @zdz2101 , please let me know when you update the FL codes.
3. One additional comment, the FL code does not seem to derive the extended BMI SDS and percentile for BMI > 95%, per the formula in the guidance: https://www.cdc.gov/growthcharts/extended-bmi-data-files.htm (see Test 5). @zdz2101 , could you look into it and let me know if there is any misunderstanding from my end?

rossfarrugia commented:
@zdz2101 please can you re-review this and if you agree with the suggested function fixes from Laura then please make a commit in this branch and we can fix and test in one place.

zdz2101 commented:
``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(rlang)
devtools::install_github("pharmaverse/admiralpeds", ref = "51-documentation-increase-unit-test-coverage-for-derive_params_growth_age")
#> Downloading GitHub repo pharmaverse/admiralpeds@51-documentation-increase-unit-test-coverage-for-derive_params_growth_age
#> 
#> ── R CMD build ─────────────────────────────────────────────────────────────────
#> * checking for file ‘/tmp/RtmpU311IW/remotes17d76f4de89db8/pharmaverse-admiralpeds-c8bb9a5/DESCRIPTION’ ... OK
#> * preparing ‘admiralpeds’:
#> * checking DESCRIPTION meta-information ... OK
#> * checking for LF line-endings in source and make files and shell scripts
#> * checking for empty or unneeded directories
#> * building ‘admiralpeds_0.1.0.9025.tar.gz’
#> Warning: invalid uid value replaced by that for user 'nobody'
#> Warning: invalid gid value replaced by that for user 'nobody'
#> Installing package into '/home/zelos.zhu/R/x86_64-pc-linux-gnu-library/4.3'
#> (as 'lib' is unspecified)
library(admiralpeds)

vs_data <- tibble::tribble(
  ~STUDYID, ~USUBJID, ~VISIT, ~SEX, ~AGECUR, ~AGEU, ~VSTESTCD, ~VSSTRESN,
  "Study", "1001", "Screening", "F", 196.5, "months", "BMI", 58.4,
)
  
meta <- cdc_bmiage %>% mutate(AGEU = "months", SEX = ifelse(SEX == 1, "M", "F"))

actual <- derive_params_growth_age(
  dataset = vs_data,
  by_vars = exprs(STUDYID, USUBJID, VISIT),
  sex = SEX,
  age = AGECUR,
  age_unit = AGEU,
  meta_criteria = meta,
  parameter = VSTESTCD == "BMI",
  bmi_cdc_correction = TRUE,
  analysis_var = VSSTRESN,
  set_values_to_sds = exprs(
    PARAMCD = "BMIASDS"
  ),
  set_values_to_pctl = exprs(
    PARAMCD = "BMIAPCTL"
  )
)
actual
#> # A tibble: 3 × 10
#>   STUDYID USUBJID VISIT     SEX   AGECUR AGEU   VSTESTCD VSSTRESN   AVAL PARAMCD
#>   <chr>   <chr>   <chr>     <chr>  <dbl> <chr>  <chr>       <dbl>  <dbl> <chr>  
#> 1 Study   1001    Screening F       196. months BMI          58.4  NA    <NA>   
#> 2 Study   1001    Screening F       196. months BMI          58.4   4.89 BMIASDS
#> 3 Study   1001    Screening F       196. months BMI          58.4 100.   BMIAPC…
```

<sup>Created on 2024-07-16 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>


@CDC-DNPAO The package version within this branch contains the proper fix to your extended BMI inquiry, there was indeed a mistake found that should be addressed now, if we haven't merged it into `main` before any other inquires you have, you can access this version of the package using 

`  devtools::install_github("pharmaverse/admiralpeds", ref = "51-documentation-increase-unit-test-coverage-for-derive_params_growth_age")`

zdz2101 reviewed (commented):
Added the snippet about new right-skewness adjusments @rossfarrugia

Minlei0201 commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -120,7 +120,7 @@ test_that("derive_params_growth_age Test 2: Height SDS and percentile works (P50
 
 ## Test 3: BMI SDS and percentile works (Z-score of -2, 0, 2, 5) ----
-test_that("derive_params_growth_age Test 3: BMI SDS and percentile works (Z-score of -2, 0, 2, 5)", {
+test_that("derive_params_growth_age Test 3: BMI SDS and percentile works (Z-score of -2, 0, 2, 5)", { # nolint
```
Hi @zdz2101 , what is the purpose of adding "# nolint" in this line?

zdz2101 replied:
Our lintr doesn't line code with more than 100 characters, so to get around it occasionally you can throw a `# nolint`

Minlei0201 commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,49 +1,427 @@
+  )
+
+  meta <- tibble::tribble(
+    ~SEX, ~AGE, ~AGEU, ~L, ~M, ~S, ~P95, ~Sigma,
```
Cool, thanks @zdz2101 for catching!

Minlei0201 commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,49 +1,427 @@
+    age_unit = AGEU,
+    meta_criteria = meta,
+    parameter = VSTESTCD == "BMI",
+    bmi_cdc_correction = TRUE,
```
Ah, yes! Thanks @zdz2101 for the update!

Minlei0201 commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,49 +1,427 @@
+    "Study", "1001", "Screening", "M", 250, "months", "WEIGHT", 58,
+  )
+
+  meta <- tibble::tribble(
```
Sounds good, thanks @zdz2101 for the update!

Minlei0201 commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,49 +1,427 @@
+    "Study", "1001", "Screening", "M", 210, "months", "WEIGHT", NA,
+  )
+
+  meta <- tibble::tribble(
```
Sounds good, thanks Zelos for the update!

Minlei0201 commented:
> > @zdz2101 please do add anything, but my understanding of the 2 questions is:
> > 
> > 1. No age conversion happens in the function. If you supply age in months in the data then you should also supply age in months in the metadata. In our template and vignette we make the metadata consistently in days via conversion and interpolation for CDC, and then we derive current age in days. Some other company could choose to only supply metadata in months and therefore derive current age in months, and the function should still work (always matching to the nearest when falling in between).
> > 2. I would agree to have both parameters created as missing would make sense to me. Maybe this is an update needed to the function?
> 
> Hi @rossfarrugia , thanks for your clarification!
> 
> 1. I've changed the age unit in days for the majority of tests, and added one test case for age in months.
> 2. @zdz2101 , please let me know when you update the FL codes.
> 3. One additional comment, the FL code does not seem to derive the extended BMI SDS and percentile for BMI > 95%, per the formula in the guidance: https://www.cdc.gov/growthcharts/extended-bmi-data-files.htm (see Test 5). @zdz2101 , could you look into it and let me know if there is any misunderstanding from my end?

Hi @zdz2101 , any update on number 2 above, regarding creating separate SDS and percentile records (eg, for out of bound age, or missing anthropometric values)?

Minlei0201 commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,49 +1,471 @@
+  expected_sds <- 3 + (20.4 - SD3pos)/(SD3pos - SD2pos)
+  expected_pctl <- pnorm(expected_sds)*100
+
+  expect_equal(
```
Hi @zdz2101 , can you remind me which email you were referring to? I searched my inbox, but didn't find an email from David on right skewness adjustment.

CDC-DNPAO replied:
Ross and Zelos - I can't tell, but be sure to do this extrapolation for both high (z >3) and low (z < (-3).

Minlei0201 replied:
Hi @rossfarrugia , no worries at all! I'm totally fine (and actually think it is a good idea) as long as it is more efficient to move forward. 
@zdz2101 , in terms of the test, the current test for weight looks good. Do we need additional test to make sure the adjustment does not apply to height and head circumference?
@CDC-DNPAO , David, I have one follow-up question on the adjustment: since it is brought up in the WHO guidance, does it only apply to the derivation with WHO metadata (i.e., patients < 2 years old)? Or is it generalizable to derivation with CDC metadata as well?

rossfarrugia replied:
thanks @Minlei0201!

I wouldn't worry to add that extra test as it's up to the user (usually dictated by our template) whether they turn this option on or off, so no need to add extra tests with it turned off for other parameters.

Good point about only applying this to the WHO metadata though. David's email said "This WHO 'adjustment' is for all z-scores (and percentiles) based on the WHO chart that involves weight, so it should also apply to weight for length, but NOT to head circumference Z or length for age Z." - @zdz2101 any clever way you can figure to only apply this for the WHO metadata given we'd never know which metadata the user passes to the function (e.g. they could give only WHO and no CDC so don't think we can hardcode the <2 yrs of age here that we use in our template)? the only thing i can see is that WHO L values are only given to 4 decimal places, whereas CDC gives more, so we could use that but it feels crude 🤷 or we have to add another column to the metadata `SOURCE` = `"CDC"` or `"WHO"`?

CDC-DNPAO replied:
@Minlei0201 @zdz2101 @rossfarrugia .  The SD32 adjustment only applies to the WHO weight-based variables.  There is no comparable SD23 adjustment for the CDC data.

rossfarrugia replied:
thanks for confirming David! now we just need to figure out a nice way to achieve that in our package... i'm beginning to see why you created 2 separate packages now... 😉

CDC-DNPAO replied:
@rossfarrugia Ross, honestly, it was for historical reasons. I needed an R CDC package when CDC only had a SAS program (which I had rewritten in ~2005).  A few years later, my uni-based friend asked me to create an R WHO package as a favor.

rossfarrugia commented on R/derive_params_growth_age.R:
```diff
@@ -72,6 +72,12 @@
+#' @param right_skew_correction Right skew correction
+#'
+#'  A logical scalar, e.g. `TRUE`/`FALSE` is expected.
+#'  WHO developed a modification to the z-score calculation to accomodate for right-skewness
```
See comment above

CDC-DNPAO commented:
Minlei - I stink at leaving comments on github.  Could you send me your
email, and I'll make sure that I include you when I send Ross and Zelos an
email.

thanks, david

On Tue, Jul 16, 2024 at 9:59 PM Minlei0201 ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In tests/testthat/test-derive_params_growth_age.R
> <https://github.com/pharmaverse/admiralpeds/pull/67#discussion_r1680304571>
> :
>
> > +    set_values_to_sds = exprs(
> +      PARAMCD = "WGASDS",
> +      PARAM = "Weight-for-age z-score"
> +    ),
> +    set_values_to_pctl = exprs(
> +      PARAMCD = "WGAPCTL",
> +      PARAM = "Weight-for-age percentile"
> +    )
> +  )
> +
> +  SD2pos <- (9.0342 * (1 + 2*0.0868*0.10885) ^ (1/0.0868))
> +  SD3pos <- (9.0342 * (1 + 3*0.0868*0.10885) ^ (1/0.0868))
> +  expected_sds <- 3 + (20.4 - SD3pos)/(SD3pos - SD2pos)
> +  expected_pctl <- pnorm(expected_sds)*100
> +
> +  expect_equal(
>
> Hi @zdz2101 <https://github.com/zdz2101> , can you remind me which email
> you were referring to? I searched my inbox, but didn't find an email from
> David on right skewness adjustment.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pharmaverse/admiralpeds/pull/67#discussion_r1680304571>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AY2XGPSEKA4GZJ5MMLNOIG3ZMXFZ3AVCNFSM6AAAAABKQD7VICVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCOBRG4YTQMJZHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Minlei0201 commented:
> Minlei - I stink at leaving comments on github. Could you send me your email, and I'll make sure that I include you when I send Ross and Zelos an email. thanks, david
> […](#)
> On Tue, Jul 16, 2024 at 9:59 PM Minlei0201 ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In tests/testthat/test-derive_params_growth_age.R <[#67 (comment)](https://github.com/pharmaverse/admiralpeds/pull/67#discussion_r1680304571)> : > + set_values_to_sds = exprs( + PARAMCD = "WGASDS", + PARAM = "Weight-for-age z-score" + ), + set_values_to_pctl = exprs( + PARAMCD = "WGAPCTL", + PARAM = "Weight-for-age percentile" + ) + ) + + SD2pos <- (9.0342 * (1 + 2*0.0868*0.10885) ^ (1/0.0868)) + SD3pos <- (9.0342 * (1 + 3*0.0868*0.10885) ^ (1/0.0868)) + expected_sds <- 3 + (20.4 - SD3pos)/(SD3pos - SD2pos) + expected_pctl <- pnorm(expected_sds)*100 + + expect_equal( Hi @zdz2101 <https://github.com/zdz2101> , can you remind me which email you were referring to? I searched my inbox, but didn't find an email from David on right skewness adjustment. — Reply to this email directly, view it on GitHub <[#67 (comment)](https://github.com/pharmaverse/admiralpeds/pull/67#discussion_r1680304571)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AY2XGPSEKA4GZJ5MMLNOIG3ZMXFZ3AVCNFSM6AAAAABKQD7VICVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCOBRG4YTQMJZHE> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Hi @CDC-DNPAO , thank you! Ross has forwarded me your email - I'll review the codes based on your email later today.

zdz2101 commented on tests/testthat/test-derive_params_growth_age.R:
```diff
@@ -1,49 +1,478 @@
+  vs_data <- tibble::tribble(
+    ~STUDYID, ~USUBJID, ~VISIT, ~SEX, ~AGECUR, ~AGEU, ~VSTESTCD, ~VSSTRESN,
+    "Study", "1001", "Screening", "M", 1233, "days", "BMI", 19,
+    "Study", "1002", "Screening", "M", 1233, "days", "BMI", 100,
```
@rossfarrugia this is the extended bmi example with the correction in place, including the setting of Z-scores to 8.21 when the percentile is extremely high

zdz2101 commented on inst/templates/ad_advs.R:
```diff
@@ -367,6 +367,7 @@ advs_ht_lgth <- advs %>%
     meta_criteria = who_wt_for_lgth,
     parameter = VSTESTCD == "WEIGHT",
     analysis_var = AVAL,
+    who_correction = TRUE,
```
all WEIGHT references now have the WHO correction turned on

rossfarrugia replied:
@zdz2101 - the WHO guideline page 302 says: "Centiles and z-scores for weight-for-age, weight-for-length, weight-for-height and **BMI-for-age**" - so we also need to add this to the BMI-for-age function call

@CDC-DNPAO alluded to this in his email when he said "based on the WHO chart that involves weight" - and BMI involves weight.

rossfarrugia replied:
Zelos - i'll make a commit to update this (including in the vignette - with an extra explanation paragraph) and please check you agree

zdz2101 reviewed (commented):
Everything should be addressed now: 

1. extended BMI reflects correctly now
2. even high/outlier values for BMI
3. the WHO correction is implemented
4. the correction is set to be `TRUE` in vignette and template for all weight references

rossfarrugia reviewed (changes requested):
@zdz2101 some review comments for the updates

rossfarrugia commented on R/derive_params_growth_age.R:
```diff
@@ -72,6 +72,12 @@
+#'
+#'  A logical scalar, e.g. `TRUE`/`FALSE` is expected.
+#'  WHO developed a modification to the z-score calculation to accommodate for right-skewness
+#'  in certain data, if set to `TRUE` the WHO correction is applied.
```
I would replace line 78 and 79 text with the following to be more in line with the explanation from @CDC-DNPAO :

According to WHO guidelines for any weight-based indicator, to overcome right skewness in the data WHO decided to not rely on usual LMS method for extreme values and instead an extapolation approach is recommended.
More details on these exact rules applied can be found at the document page 302 of the [WHO Child Growth Standards Guidelines] (https://www.who.int/publications/i/item/924154693X).
If set to `TRUE` the WHO correction is applied.

rossfarrugia replied:
Zelos - i'll make a commit to update this and please check you agree

CDC-DNPAO replied:
@zdz2101 @rossfarrugia.  This SD23 WHO adjustment doesn't have anything to do with right skewness.  Actually, I can't even find 'right skewness' in the WHO 300+ page document.  Honestly, WHO did this because they decided to delete all kids with an abs z-score >3.  Then, they had to do something to calculate these extreme z-scores.  I would just refer to this as the  'WHO adjustment' or the WHO SD23 adjustment'.  

It's the L parameter in LMS that adjusts for right skewness using the Box-Cox transformation.   If I ever used the term 'right adjustment' for this, I apologize for my mistake

zdz2101 replied:
I saw the phrase right-skewness somewhere on pg 302 before the formula, 
![image](https://github.com/user-attachments/assets/4948f9a3-bf0e-4a1c-8f3f-5015749f5878)


I'll remove any instances of the word "right skewnesss" and simply call it the `WHO adjustment` I don't think it's necessary for users to know the history of the decision made

CDC-DNPAO replied:
I agree that 'WHO adjustment' is much better

rossfarrugia commented on R/derive_params_growth_height.R:
```diff
@@ -68,7 +68,13 @@
+#'
+#'  A logical scalar, e.g. `TRUE`/`FALSE` is expected.
+#'  WHO developed a modification to the z-score calculation to accommodate for right-skewness
+#'  in certain data, if set to `TRUE` the WHO correction is applied.
```
See comment above

rossfarrugia replied:
Zelos - i'll make a commit to update this and please check you agree

CDC-DNPAO replied:
@rossfarrugia @zdz2101 @milei.  About WHO BMI SDS and pctls.  WHO does calculate these, but almost all people use weight-for-length rather than BMI among <2s.  The only times I've seen BMI used among infants is when someone wants to see if BMI SDS is a better predictor of subsequent obesity than is WFL SDs, and the results are not strong enough to change to BMI SDS.  The only advantage of BMI SDS that I know of is that it isn't limited by the length of the infants.  For WFL SDS, if you've got a very short or very tall child, the length will fall outside the WHO length limits and  WFL z/pctl will be NA.  If it isn't difficult to add BMI SDS/pctl for infants to the program, go ahead, but honestly, I don't think it's a priority because it's so little used.

CDC-DNPAO replied:
@Ross - sorry, I had no idea that Roche wanted this.  I think this is a really strange request, but glad it won't be extra work.

rossfarrugia replied:
no problem! i might query it with them later when i come to build the Roche specific template based on admiralpeds, as there's a lot of other things missed too from their standards like the WHO skewness piece you pointed out to us.

rossfarrugia reviewed (changes requested):
@zdz2101 a few more updates spotted

rossfarrugia commented on R/derive_params_growth_age.R:
```diff
@@ -276,14 +291,26 @@ derive_params_growth_age <- function(dataset,
+        mutate(
+          AVAL := case_when( # nolint
+            AVAL > 3 ~ 3 + ({{ analysis_var }} - SD3pos) / (SD3pos - SD2pos),
+            AVAL < -3 ~ -3 + ({{ analysis_var }} - SD3neg) / (SD2neg - SD3neg),
```
Zelos - i'll make a commit to update this and please check you agree

Minlei0201 replied:
Hi @rossfarrugia , from WHO guidance p302, if zind*<-3, the formula is -3**+**[(y-SD3neg)/(SD2neg-SD3pos)]. In David's email, he also has abs in the formula: z < (-3), -3 - **abs**((var - sdm3)/(sdm2 - sdm3)). So I think if we don't use abs, Zelos should be using the correct formula?

rossfarrugia commented on R/derive_params_growth_age.R:
```diff
@@ -299,6 +326,17 @@ derive_params_growth_age <- function(dataset,
+        mutate(
+          AVAL := case_when( # nolint
+            AVAL > 3 ~ 3 + ({{ analysis_var }} - SD3pos) / (SD3pos - SD2pos),
+            AVAL < -3 ~ -3 + ({{ analysis_var }} - SD3neg) / (SD2neg - SD3neg),
```
Zelos - i'll make a commit to update this and please check you agree

CDC-DNPAO replied:
@rossfarrugia  @zdz2101   @Minlei0201 I think Zelos's original formula, with a '+' for values <  -3, really parallels the WHO formulas and is correct.  'var - SD3neg' is always going to be negative, while SD2neg-SD3neg is always going to be positive.  So you want the 2nd part of the formula, which will be negative, to be added (not subtracted) from -3.  

This brings back the frustrating stuff I did 15 years ago - I had trouble with the formula, but when I finally found that using 'abs' resulted in the correct answer, I called it a day!

rossfarrugia commented on R/derive_params_growth_height.R:
```diff
@@ -303,6 +319,17 @@ derive_params_growth_height <- function(dataset,
+        mutate(
+          AVAL := case_when( # nolint
+            AVAL > 3 ~ 3 + ({{ analysis_var }} - SD3pos) / (SD3pos - SD2pos),
+            AVAL < -3 ~ -3 + ({{ analysis_var }} - SD3neg) / (SD2neg - SD3neg),
```
Zelos - i'll make a commit to update this and please check you agree

rossfarrugia commented on tests/testthat/test-derive_params_growth_height.R:
```diff
@@ -236,3 +238,52 @@ test_that("derive_params_growth_height Test 5: returns expected error message",
+  sd3neg <- (3.3278 * (1 - 3 * -0.3521 * 0.08890)^(1 / -0.3521))
+  expected_sds <- c(
+    3 + (5 - sd3pos) / (sd3pos - sd2pos),
+    -3 + (2 - sd3neg) / (sd2neg - sd3neg)
```
Zelos - i'll make a commit to update this and please check you agree

rossfarrugia commented on R/derive_params_growth_age.R:
```diff
@@ -276,14 +291,26 @@ derive_params_growth_age <- function(dataset,
             AVAL
-          )
+          ),
+          AVAL = ifelse(AVAL == Inf, 8.21, AVAL)
```
Zelos - i'll make a commit to update this and please check you agree

rossfarrugia commented on R/derive_params_growth_age.R:
```diff
@@ -307,7 +345,7 @@ derive_params_growth_age <- function(dataset,
             AVAL
           ),
-          AVAL = pnorm(AVAL) * 100
+          AVAL = ifelse(AVAL == Inf, 8.21, AVAL)
```
Zelos - i'll make a commit to update this and please check you agree

rossfarrugia commented:
@zdz2101 i made a commit to fix a number of the remaining comments - please double check.

from what i can make of it now we just have 2 things left to fix here:

- the comment from @Minlei0201 way above in this PR: For out of bound age and/or missing anthropometric measures, if the user requests both SDS and percentile, to my understanding, the function is supposed to return two records with missing values (i.e., one for SDS and one for percentile). However, currently the function only returns one record with missing value (see Test 6 and 7). Is that what you expect?

- How to only apply the `who_correction` option for ages when using WHO metadata (not CDC)? See discussion above 🤷 🤯

CDC-DNPAO commented:
These L and S values are not hard-coded into the program, are they? I'm
hoping this is simply an example for one child. When you calculate SD2,
SD2, etc., you've got to use the WHO L and S values for a child of that
specific sex and age.

On Thu, Jul 18, 2024 at 4:17 AM Ross Farrugia ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In tests/testthat/test-derive_params_growth_height.R
> <https://github.com/pharmaverse/admiralpeds/pull/67#discussion_r1682414355>
> :
>
> > +      PARAMCD = "WGHSDS",
> +      PARAM = "Weight-for-height Z-Score"
> +    ),
> +    set_values_to_pctl = exprs(
> +      PARAMCD = "WGHPCTL",
> +      PARAM = "Weight-for-height percentile"
> +    )
> +  )
> +
> +  sd2pos <- (3.3278 * (1 + 2 * -0.3521 * 0.08890)^(1 / -0.3521))
> +  sd3pos <- (3.3278 * (1 + 3 * -0.3521 * 0.08890)^(1 / -0.3521))
> +  sd2neg <- (3.3278 * (1 - 2 * -0.3521 * 0.08890)^(1 / -0.3521))
> +  sd3neg <- (3.3278 * (1 - 3 * -0.3521 * 0.08890)^(1 / -0.3521))
> +  expected_sds <- c(
> +    3 + (5 - sd3pos) / (sd3pos - sd2pos),
> +    -3 + (2 - sd3neg) / (sd2neg - sd3neg)
>
> Zelos - i'll make a commit to update this and please check you agree
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pharmaverse/admiralpeds/pull/67#discussion_r1682414355>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AY2XGPUO3VPDZDET7KSM5I3ZM52Z5AVCNFSM6AAAAABKQD7VICVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCOBVGA3TAMZQGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rossfarrugia commented:
> These L and S values are not hard-coded into the program, are they? I'm hoping this is simply an example for one child. When you calculate SD2, SD2, etc., you've got to use the WHO L and S values for a child of that specific sex and age.

@CDC-DNPAO that one's a unit test program. It's not part of the function code, it is just used to test the results are reliable, so we feed in test data values and you check that the expected outcomes match what we expect.

zdz2101 commented on R/derive_params_growth_age.R:
```diff
@@ -296,7 +300,7 @@ derive_params_growth_age <- function(dataset,
           AVAL := case_when( # nolint
             AVAL > 3 ~ 3 + ({{ analysis_var }} - SD3pos) / (SD3pos - SD2pos),
-            AVAL < -3 ~ -3 + ({{ analysis_var }} - SD3neg) / (SD2neg - SD3neg),
+            AVAL < -3 ~ -3 - ({{ analysis_var }} - SD3neg) / (SD2neg - SD3neg),
```
![image](https://github.com/user-attachments/assets/0c7b3b41-91d7-42af-bd88-dd1e9f9549fc)

It was correct as is, I followed the formulation directly from the textbook, David uses 
`z < (-3), -3 - abs((var - sdm3)/(sdm2 - sdm3))`

which is the same as long as the right side fractional term is negative

CDC-DNPAO replied:
@zdz2101 I like your style more than what I used!

CDC-DNPAO commented:
Sorry, I omitted the hyphen in my search of the WHO document.

But the sentences following your png PDF say:

'LMS method fits skewed data adequately by using a Box-Cox normal
distribution, which follows the empirical data closely. The drawback,
however, is that the outer tails of the distribution are highly affected by
extreme data points even if only very few (e.g. less than 1%). A restricted
application of the LMS method was thus used for the construction of the WHO
weight-based indicators, limiting the Box-Cox normal distribution to the
interval corresponding to z-scores where empirical data were available
(i.e. between -3 SD and 3 SD).'

I would change your 'account for right-skewness' to 'who_adjustment
weight-based calculations'.  I agree that users don't need to know my
version of the history of this.

david

On Fri, Jul 19, 2024 at 10:43 AM Zelos Zhu ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In R/derive_params_growth_age.R
> <https://github.com/pharmaverse/admiralpeds/pull/67#discussion_r1684490143>
> :
>
> > @@ -72,6 +72,12 @@
>  #'  CDC developed extended percentiles (>95%) to monitor high BMI values,
>  #'  if set to `TRUE` the CDC's correction is applied.
>  #'
> +#' @param who_correction Right skew correction
> +#'
> +#'  A logical scalar, e.g. `TRUE`/`FALSE` is expected.
> +#'  WHO developed a modification to the z-score calculation to accommodate for right-skewness
> +#'  in certain data, if set to `TRUE` the WHO correction is applied.
>
> I saw the phrase right-skewness somewhere on pg 302 before the formula,
> image.png (view on web)
> <https://github.com/user-attachments/assets/4948f9a3-bf0e-4a1c-8f3f-5015749f5878>
>
> I'll remove any instances of the word "right skewnesss" and simply call it
> the "WHO adjustment" I don't think it's necessary for users to know the
> history of the decision made
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pharmaverse/admiralpeds/pull/67#discussion_r1684490143>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AY2XGPUZ6UGAFVLL63WADDLZNEQZZAVCNFSM6AAAAABKQD7VICVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCOBYGQYTKNZZG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

zdz2101 reviewed (approved):
@rossfarrugia Your updates look good to me!
